interop/xds/client/client.go

```diff
@@ -0,0 +1,174 @@
+ * limitations under the License.
+ *
+ */
+
+// Binary client for xDS interop tests.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/peer"
+)
+
+type statsWatcherKey struct {
+	startID int32
+	endID   int32
+}
+
+type statsWatcher struct {
+	rpcsByPeer    map[string]int32
+	numFailures   int32
+	remainingRpcs int32
+	c             chan testpb.SimpleResponse
+}
+
+var (
+	numChannels   = flag.Int("num_channels", 1, "Num of channels")
+	printResponse = flag.Bool("print_response", false, "Write RPC response to stdout")
+	qps           = flag.Int("qps", 1, "QPS per channel")
+	rpcTimeout    = flag.Duration("rpc_timeout", 10*time.Second, "Per RPC timeout")
+	server        = flag.String("server", "localhost:8080", "Address of server to connect to")
+	statsPort     = flag.Int("stats_port", 8081, "Port to expose peer distribution stats service")
+
+	mu               sync.Mutex
+	currentRequestID int32
+	watchers         = make(map[statsWatcherKey]*statsWatcher)
+)
+
+type statsService struct{}
+
+// Wait for the next LoadBalancerStatsRequest.GetNumRpcs to start and complete,
+// and return the distribution of remote peers. This is essentially a clientside
+// LB reporting mechanism that is designed to be queried by an external test
+// driver when verifying that the client is distributing RPCs as expected.
+func (s *statsService) GetClientStats(ctx context.Context, in *testpb.LoadBalancerStatsRequest) (*testpb.LoadBalancerStatsResponse, error) {
+	mu.Lock()
+	watcherKey := statsWatcherKey{currentRequestID, currentRequestID + in.GetNumRpcs()}
+	watcher, ok := watchers[watcherKey]
+	if !ok {
+		watcher = &statsWatcher{
+			rpcsByPeer:    make(map[string]int32),
+			numFailures:   0,
+			remainingRpcs: in.GetNumRpcs(),
+			c:             make(chan testpb.SimpleResponse),
+		}
+		watchers[watcherKey] = watcher
+	}
+	mu.Unlock()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(in.GetTimeoutSec())*time.Second)
+	defer cancel()
+
+	responseCh := make(chan testpb.LoadBalancerStatsResponse)
+	// Wait until the requested RPCs have all been recorded or timeout occurs.
+	go func() {
+		for {
+			select {
+			case r := <-watcher.c:
+				if r.GetServerId() != "" {
+					watcher.rpcsByPeer[r.GetServerId()]++
+				} else {
+					watcher.numFailures++
+				}
+				watcher.remainingRpcs--
+				if watcher.remainingRpcs == 0 {
+					responseCh <- testpb.LoadBalancerStatsResponse{NumFailures: watcher.numFailures + watcher.remainingRpcs, RpcsByPeer: watcher.rpcsByPeer}
+					return
+				}
+			case <-ctx.Done():
+				grpclog.Info("Timed out, returning partial stats")
+				responseCh <- testpb.LoadBalancerStatsResponse{NumFailures: watcher.numFailures + watcher.remainingRpcs, RpcsByPeer: watcher.rpcsByPeer}
+				return
+			}
+		}
+	}()
+
+	response := <-responseCh
```

Comment:
This is one goroutine block waiting for another one goroutine. This channel isn't necessary. Just inline the `for select` loop. When the loop breaks (either all RPCs finish, or timeout), the response is ready.

Reply:
Done